### PR TITLE
tests: adjust stack size for mps2_an385's coverage test

### PR DESCRIPTION
**Pull request by wentongwu**
for SDK 0.10.0, it consumes more stack size when coverage enabled
on mps2_an385 platform, adjust stack size for most of the test cases,
otherwise there will be stack overflow.

Fixes: #14500

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

### CODE REVIEW

File: tests/kernel/critical/src/main.c [hunk 1]
```diff
@@ -42,7 +42,12 @@ static struct k_work_q offload_work_q;
 static K_THREAD_STACK_DEFINE(offload_work_q_stack,
 			     CONFIG_OFFLOAD_WORKQUEUE_STACK_SIZE);
 
+#ifdef CONFIG_COVERAGE_GCOV
+#define STACK_SIZE 4096
+#else
 #define STACK_SIZE 1024
```
Comment (andrewboie) [inline tests/kernel/critical/src/main.c]: just add CONFIG_TEST_EXTRA_STACKSIZE instead

File: tests/kernel/workq/work_queue/src/main.c [hunk 2]
```diff
@@ -25,7 +25,11 @@
  */
 #define SUBMIT_WAIT             50
 
+#ifdef CONFIG_COVERAGE_GCOV
+#define STACK_SIZE      4096
+#else
 #define STACK_SIZE      1024
```
Comment (andrewboie) [inline tests/kernel/workq/work_queue/src/main.c]: just add CONFIG_TEST_EXTRA_STACKSIZE instead

File: tests/kernel/threads/thread_apis/src/test_essential_thread.c [hunk 3]
```diff
@@ -10,7 +10,11 @@
 
 struct k_thread kthread_thread;
 
+#ifdef CONFIG_COVERAGE_GCOV
+#define STACKSIZE 4096
+#else
 #define STACKSIZE 1024
```
Comment (andrewboie) [inline tests/kernel/threads/thread_apis/src/test_essential_thread.c]: just add CONFIG_TEST_EXTRA_STACKSIZE instead

File: tests/kernel/sched/preempt/src/main.c [hunk 4]
```diff
@@ -49,7 +49,7 @@ const enum { METAIRQ, COOP, PREEMPTIBLE } worker_priorities[] = {
 #define NUM_THREADS ARRAY_SIZE(worker_priorities)
 
 #ifdef CONFIG_COVERAGE
-#define STACK_SIZE (512 + CONFIG_TEST_EXTRA_STACKSIZE)
+#define STACK_SIZE 4096
```
Comment (andrewboie) [inline tests/kernel/sched/preempt/src/main.c]: This shouldn't need to change?

File: tests/kernel/poll/src/test_poll.c [hunk 5]
```diff
@@ -526,7 +538,11 @@ void test_poll_multi(void)
 }
 
 static struct k_thread signal_thread;
+#ifdef CONFIG_COVERAGE_GCOV
+static K_THREAD_STACK_DEFINE(signal_stack, KB(4));
+#else
 static K_THREAD_STACK_DEFINE(signal_stack, KB(1));
```
Comment (andrewboie) [inline tests/kernel/poll/src/test_poll.c]: just add CONFIG_TEST_EXTRA_STACKSIZE instead

File: tests/kernel/poll/src/test_poll.c [hunk 6]
```diff
@@ -449,7 +457,11 @@ static void multi_lowprio(void *p1, void *p2, void *p3)
 static K_SEM_DEFINE(multi_reply, 0, 1);
 
 static struct k_thread multi_thread;
+#ifdef CONFIG_COVERAGE_GCOV
+static K_THREAD_STACK_DEFINE(multi_stack, KB(4));
+#else
 static K_THREAD_STACK_DEFINE(multi_stack, KB(1));
```
Comment (andrewboie) [inline tests/kernel/poll/src/test_poll.c]: just add CONFIG_TEST_EXTRA_STACKSIZE instead

File: tests/kernel/poll/src/test_poll.c [hunk 7]
```diff
@@ -429,7 +433,11 @@ void test_poll_cancel_main_high_prio(void)
 static K_SEM_DEFINE(multi_sem, 0, 1);
 
 static struct k_thread multi_thread_lowprio;
+#ifdef CONFIG_COVERAGE_GCOV
+static K_THREAD_STACK_DEFINE(multi_stack_lowprio, KB(4));
+#else
 static K_THREAD_STACK_DEFINE(multi_stack_lowprio, KB(1));
```
Comment (andrewboie) [inline tests/kernel/poll/src/test_poll.c]: just add CONFIG_TEST_EXTRA_STACKSIZE instead

File: tests/kernel/poll/src/test_poll.c [hunk 8]
```diff
@@ -114,7 +114,11 @@ static struct k_poll_signal wait_signal =
 struct fifo_msg wait_msg = { NULL, FIFO_MSG_VALUE };
 
 static struct k_thread poll_wait_helper_thread;
+#ifdef CONFIG_COVERAGE_GCOV
+static K_THREAD_STACK_DEFINE(poll_wait_helper_stack, KB(4));
+#else
 static K_THREAD_STACK_DEFINE(poll_wait_helper_stack, KB(1));
```
Comment (andrewboie) [inline tests/kernel/poll/src/test_poll.c]: just add CONFIG_TEST_EXTRA_STACKSIZE instead

File: tests/kernel/pipe/pipe_api/src/test_pipe_contexts.c [hunk 9]
```diff
@@ -6,7 +6,11 @@
 
 #include <ztest.h>
 
+#ifdef CONFIG_COVERAGE_GCOV
+#define STACK_SIZE	4096
+#else
 #define STACK_SIZE	1024
```
Comment (andrewboie) [inline tests/kernel/pipe/pipe_api/src/test_pipe_contexts.c]: just add CONFIG_TEST_EXTRA_STACKSIZE instead

File: tests/kernel/obj_tracing/src/main.c [hunk 10]
```diff
@@ -10,7 +10,11 @@
 
 extern void test_obj_tracing(void);
 
+#ifdef CONFIG_COVERAGE_GCOV
+#define STSIZE 4096
+#else
 #define STSIZE 1024
```
Comment (andrewboie) [inline tests/kernel/obj_tracing/src/main.c]: just add CONFIG_TEST_EXTRA_STACKSIZE instead

File: tests/kernel/lifo/lifo_usage/src/main.c [hunk 11]
```diff
@@ -64,7 +68,11 @@ struct timeout_order_data timeout_order_data_mult_lifo[] = {
 
 #define NUM_SCRATCH_LIFO_PACKETS 20
 #define TIMEOUT_ORDER_NUM_THREADS	ARRAY_SIZE(timeout_order_data_mult_lifo)
+#ifdef CONFIG_COVERAGE_GCOV
+#define TSTACK_SIZE			4096
+#else
 #define TSTACK_SIZE			1024
```
Comment (andrewboie) [inline tests/kernel/lifo/lifo_usage/src/main.c]: just add CONFIG_TEST_EXTRA_STACKSIZE instead

File: tests/kernel/lifo/lifo_usage/src/main.c [hunk 12]
```diff
@@ -8,7 +8,11 @@
 #include "lifo_usage.h"
 #include <kernel.h>
 
+#ifdef CONFIG_COVERAGE_GCOV
+#define STACK_SIZE 4096
+#else
 #define STACK_SIZE 1024
```
Comment (andrewboie) [inline tests/kernel/lifo/lifo_usage/src/main.c]: just add CONFIG_TEST_EXTRA_STACKSIZE instead

File: tests/kernel/fifo/fifo_usage/src/main.c [hunk 13]
```diff
@@ -35,7 +35,11 @@
 #include <ztest.h>
 #include <irq_offload.h>
 
+#ifdef CONFIG_COVERAGE_GCOV
+#define STACK_SIZE	4096
+#else
 #define STACK_SIZE	1024
```
Comment (andrewboie) [inline tests/kernel/fifo/fifo_usage/src/main.c]: just add CONFIG_TEST_EXTRA_STACKSIZE instead

File: tests/kernel/fifo/fifo_timeout/src/main.c [hunk 14]
```diff
@@ -76,7 +76,11 @@ struct timeout_order_data timeout_order_data_mult_fifo[] = {
 };
 
 #define TIMEOUT_ORDER_NUM_THREADS	ARRAY_SIZE(timeout_order_data_mult_fifo)
+#ifdef CONFIG_COVERAGE_GCOV
+#define TSTACK_SIZE			4096
+#else
 #define TSTACK_SIZE			1024
```
Comment (andrewboie) [inline tests/kernel/fifo/fifo_timeout/src/main.c]: just add CONFIG_TEST_EXTRA_STACKSIZE instead

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]

File: tests/kernel/sched/preempt/src/main.c (hunk 4 above)
Comment (wentongwu) [reply tests/kernel/sched/preempt/src/main.c]: In this test’s prj.conf , it defines a small CONFIG_TEST_EXTRA_STACKSIZE.

File: tests/kernel/threads/thread_apis/src/test_essential_thread.c (hunk 3 above)
Comment (wentongwu) [reply tests/kernel/threads/thread_apis/src/test_essential_thread.c]: In this test’s prj.conf , it defines a small CONFIG_TEST_EXTRA_STACKSIZE.

File: tests/kernel/critical/src/main.c (hunk 1 above)
Comment (wentongwu) [reply tests/kernel/critical/src/main.c]: Ok, I will do the change. Please help review another pr for qemu_x86, two platforms failed on different cases.

Comment (andrewboie) [reply tests/kernel/critical/src/main.c]: it's the same bug on both targets, just do in a single PR.

**Comment by wentongwu** [comment]
@andrewboie  I have update the patch as review requested.
For the shippable failed, I checked the logs, it failed two cases for nrf52810_pca10040, both are because of not enough sram space, I adjust stack size for my PR so many times, it still failed. 
So I update my local code to latest version, without my PR, it also failed with same reason, So it's not my PR cause shippable failed. 




 Failures
Sanitycheck / nrf52810_pca10040:tests/kernel/sched/preempt/kernel.sched.preempt / nrf52810_pca10040/tests/kernel/sched/preempt/kernel.sched.preempt 
FAILED: : && ccache /opt/sdk/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc    zephyr/CMakeFiles/zephyr_prebuilt.dir/misc/empty_file.c.obj  -o zephyr/zephyr_prebuilt.elf  -T zephyr/linker.cmd zephyr/priv_stacks_hash_renamed.o -Wl,-Map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/out-3nd-pass/nrf52810_pca10040/tests/kernel/sched/preempt/kernel.sched.preempt/zephyr/zephyr.map -u_OffsetAbsSyms -u_ConfigAbsSyms -Wl,--whole-archive app/libapp.a zephyr/libzephyr.a zephyr/arch/arch/arm/core/libarch__arm__core.a zephyr/arch/arch/arm/core/cortex_m/libarch__arm__core__cortex_m.a zephyr/arch/arch/arm/core/cortex_m/mpu/libarch__arm__core__cortex_m__mpu.a zephyr/lib/libc/minimal/liblib__libc__minimal.a zephyr/subsys/testsuite/ztest/libsubsys__testsuite__ztest.a zephyr/drivers/gpio/libdrivers__gpio.a zephyr/drivers/serial/libdrivers__serial.a -Wl,--no-whole-archive zephyr/kernel/libkernel.a zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj -L"/opt/sdk/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/thumb/v7e-m/nofp" -L/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/out-3nd-pass/nrf52810_pca10040/tests/kernel/sched/preempt/kernel.sched.preempt/zephyr -lgcc -Wl,--print-memory-usage -mthumb -mcpu=cortex-m4 -nostdlib -static -no-pie -Wl,--fatal-warnings -Wl,-X -Wl,-N -Wl,--gc-sections -Wl,--build-id=none -Wl,--sort-common=descending -Wl,--sort-section=alignment -Wl,--orphan-handling=warn -mabi=aapcs -march=armv7e-m && :
Memory region         Used Size  Region Size  %age Used
           FLASH:       63644 B       192 KB     32.37%
            SRAM:         25 KB        24 KB    104.17%
        IDT_LIST:          56 B         2 KB      2.73/opt/sdk/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/../../../../arm-zephyr-eabi/bin/ld: zephyr/zephyr_prebuilt.elf section `priv_stacks_noinit' will not fit in region `SRAM'
/opt/sdk/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/../../../../arm-zephyr-eabi/bin/ld: section .intList VMA [0000000020006000,0000000020006037] overlaps section priv_stacks_noinit VMA [0000000020003800,00000000200063ff]
/opt/sdk/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/../../../../arm-zephyr-eabi/bin/ld: region `SRAM' overflowed by 1024 bytes
collect2: error: ld returned 1 exit status
%
ninja: build stopped: subcommand failed.

 Sanitycheck / nrf52810_pca10040:tests/kernel/poll/kernel.poll / nrf52810_pca10040/tests/kernel/poll/kernel.poll 
[128/139] Linking C executable zephyr/zephyr_prebuilt.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       71324 B       192 KB     36.28%
            SRAM:         24 KB        24 KB    100.00%
        IDT_LIST:          56 B         2 KB      2.73%
[129/139] Generating kobject_hash.gperf
[130/139] Generating linker_pass_final.cmd
[131/139] Generating kobject_hash_preprocessed.c
[132/139] Generating kobject_hash.c
[133/139] Building C object zephyr/CMakeFiles/output_lib.dir/kobject_hash.c.obj
[134/139] Linking C static library zephyr/liboutput_lib.a
[135/139] Generating kobject_hash_renamed.o
[136/139] Generating isr_tables.c
[137/139] Building C object zephyr/CMakeFiles/kernel_elf.dir/misc/empty_file.c.obj
[138/139] Building C object zephyr/CMakeFiles/kernel_elf.dir/isr_tables.c.obj
[139/139] Linking C executable zephyr/zephyr.elf
FAILED: : && ccache /opt/sdk/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc    zephyr/CMakeFiles/kernel_elf.dir/misc/empty_file.c.obj zephyr/CMakeFiles/kernel_elf.dir/isr_tables.c.obj  -o zephyr/zephyr.elf  zephyr/priv_stacks_hash_renamed.o zephyr/kobject_hash_renamed.o -T zephyr/linker_pass_final.cmd -Wl,-Map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/out-3nd-pass/nrf52810_pca10040/tests/kernel/poll/kernel.poll/zephyr/zephyr.map -u_OffsetAbsSyms -u_ConfigAbsSyms -Wl,--whole-archive app/libapp.a zephyr/libzephyr.a zephyr/arch/arch/arm/core/libarch__arm__core.a zephyr/arch/arch/arm/core/cortex_m/libarch__arm__core__cortex_m.a zephyr/arch/arch/arm/core/cortex_m/mpu/libarch__arm__core__cortex_m__mpu.a zephyr/lib/libc/minimal/liblib__libc__minimal.a zephyr/subsys/testsuite/ztest/libsubsys__testsuite__ztest.a zephyr/drivers/gpio/libdrivers__gpio.a zephyr/drivers/serial/libdrivers__serial.a -Wl,--no-whole-archive zephyr/kernel/libkernel.a zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj -L"/opt/sdk/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/thumb/v7e-m/nofp" -L/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/out-3nd-pass/nrf52810_pca10040/tests/kernel/poll/kernel.poll/zephyr -lgcc -mthumb -mcpu=cortex-m4 -nostdlib -static -no-pie -Wl,--fatal-warnings -Wl,-X -Wl,-N -Wl,--gc-sections -Wl,--build-id=none -Wl,--sort-common=descending -Wl,--sort-section=alignment -Wl,--orphan-handling=warn -mabi=aapcs -march=armv7e-m && cd /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/out-3nd-pass/nrf52810_pca10040/tests/kernel/poll/kernel.poll/zephyr && /usr/bin/python3 /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/scripts/check_link_map.py zephyr.map && /opt/sdk/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-objcopy -S -Oihex --gap-fill 0xff -R .comment -R COMMON -R .eh_frame zephyr.elf zephyr.hex && /opt/sdk/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-objcopy -S -Obinary --gap-fill 0xff -R .comment -R COMMON -R .eh_frame zephyr.elf zephyr.bin && /opt/sdk/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-objdump -S zephyr.elf > zephyr.lst && /opt/sdk/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-readelf -e zephyr.elf > zephyr.stat
/opt/sdk/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/../../../../arm-zephyr-eabi/bin/ld: zephyr/zephyr.elf section `priv_stacks_noinit' will not fit in region `SRAM'
/opt/sdk/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/../../../../arm-zephyr-eabi/bin/ld: region `SRAM' overflowed by 2048 bytes
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.

**Comment by wentongwu** [comment]
@andrewboie could you please review the new changes? thanks

File: arch/Kconfig [hunk 15]
```diff
@@ -199,7 +199,7 @@ config PRIVILEGED_STACK_TEXT_AREA
 
 config KOBJECT_TEXT_AREA
 	int "Size if kobject text area"
-	default 512 if COVERAGE_GCOV
+	default 2048 if COVERAGE_GCOV
```
Comment (dcpleung) [inline arch/Kconfig]: Master should not need this change anymore.

**Comment by andrewboie** [comment]
We need a passing CI result

File: tests/kernel/threads/thread_apis/src/test_essential_thread.c (hunk 3 above)
Comment (andrewboie) [reply tests/kernel/threads/thread_apis/src/test_essential_thread.c]: that really shouldn't be set in prj.conf any clue why it is doing it?

File: arch/Kconfig (hunk 15 above)
Comment (wentongwu) [reply arch/Kconfig]: yes, I see your PR, thanks

**Comment by wentongwu** [comment]
> We need a passing CI result
temporary exclude nrf52810_pca10040 on that two cases, when issue #15103 fixed, open it again?

**Comment by wentongwu** [comment]
@andrewboie @dcpleung could you please review the new changes? Thanks!

File: tests/kernel/poll/testcase.yaml [hunk 16]
```diff
@@ -2,3 +2,4 @@ tests:
   kernel.poll:
     tags: kernel userspace
     min_ram: 16
+    platform_exclude: nrf52810_pca10040
```
Comment (dcpleung) [inline tests/kernel/poll/testcase.yaml]: Could you extract the two platform_exclude changes into its own commit? And explain in the commit message why this platform is excluded.

File: tests/kernel/threads/thread_apis/src/main.c [hunk 17]
```diff
@@ -37,7 +37,7 @@ extern void test_k_thread_foreach(void);
 extern void test_threads_cpu_mask(void);
 
 struct k_thread tdata;
-#define STACK_SIZE (256 + CONFIG_TEST_EXTRA_STACKSIZE)
+#define STACK_SIZE (512 + CONFIG_TEST_EXTRA_STACKSIZE)
```
Comment (dcpleung) [inline tests/kernel/threads/thread_apis/src/main.c]: CI has been passing (mostly) without this type of changes which also increases the stack size for non-coverage tests. I would recommend defining `CONFIG_TEST_EXTRA_STACKSIZE` in `prj*.conf` instead. We still want to build these with sanitycheck for platforms with smaller memory space.

**Review by dcpleung — CHANGES_REQUESTED** [review]

File: tests/kernel/threads/thread_apis/src/main.c (hunk 17 above)
Comment (wentongwu) [reply tests/kernel/threads/thread_apis/src/main.c]: actually I didn't change the original stack size. original logic is that 256 + CONFIG_TEST_EXTRA_STACKSIZE(CONFIG_TEST_EXTRA_STACKSIZE=256 in prj.conf) = 512, my change just remove the defining CONFIG_TEST_EXTRA_STACKSIZE and add every digital number defining stack size by 256.

File: tests/kernel/poll/testcase.yaml (hunk 16 above)
Comment (wentongwu) [reply tests/kernel/poll/testcase.yaml]: nrf52810_pca10040 platform will fail at these two test cases as issue #15103 described, and CI will run these two cases on that platform when this PR is submitted.

Comment (dcpleung) [reply tests/kernel/poll/testcase.yaml]: A PR can contain multiple commits and the CI will check against the whole PR (including all commits).

File: tests/kernel/threads/thread_apis/src/main.c (hunk 17 above)
Comment (dcpleung) [reply tests/kernel/threads/thread_apis/src/main.c]: Ah... I see. I missed the change in prj.conf.

File: tests/kernel/poll/testcase.yaml (hunk 16 above)
Comment (wentongwu) [reply tests/kernel/poll/testcase.yaml]: thanks,

**Comment by wentongwu** [comment]
@andrewboie @dcpleung could you please review the new patch? Thanks!

**Review by tyrantlizard77 — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Pull request merged**